Migrate rpow as LLK op

Aswinmcw opened the pull request:
### Ticket
https://github.com/tenstorrent/tt-metal/issues/24199

### Problem description
rpow implemented as composite op

### What's changed
Implement it as llk op

Edge case behaviour:

| Expression              | ttnn.rpow Result | Golden Tensor |
|--------------------------|------------------|---------------|
| ttnn.rpow(5, inf)        | inf              | inf           |
| ttnn.rpow(5, -inf)       | -inf             | -inf          |
| ttnn.rpow(5, nan)        | **inf** 🔴       | **nan**       |
| ttnn.rpow(inf, 5)        | inf              | inf           |
| ttnn.rpow(-inf, 5)       | 0                | 0             |
| ttnn.rpow(nan, 5)        | **inf** 🔴       | **nan**       |
| ttnn.rpow(5, inf)        | inf              | inf           |
| ttnn.rpow(inf, inf)      | inf              | inf           |
| ttnn.rpow(inf, -inf)     | inf              | inf           |
| ttnn.rpow(inf, nan)      | **inf** 🔴       | **nan**       |
| ttnn.rpow(-inf, inf)     | 0                | 0             |
| ttnn.rpow(-inf, -inf)    | **inf** 🔴       | **0**         |
| ttnn.rpow(-inf, nan)     | **0** 🔴         | **nan**       |
| ttnn.rpow(nan, inf)      | **inf** 🔴       | **nan**       |
| ttnn.rpow(nan, -inf)     | **inf** 🔴       | **nan**       |
| ttnn.rpow(nan, nan)      | **inf** 🔴       | **nan**       |

### Checklist
- [ ] [All post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml) CI passes
- [ ] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml) CI with demo tests passes (if applicable)
- [ ] [Model regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-models.yaml) CI passes (if applicable)
- [ ] [Device performance regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-device-models.yaml) CI passes (if applicable)
- [ ] (For models and ops writers) [Single-card demo tests](https://github.com/tenstorrent/tt-metal/actions/workflows/single-card-demo-tests.yaml) CI passes (if applicable) See [recommended dev flow](https://github.com/tenstorrent/tt-metal/blob/main/models/docs/MODEL_ADD.md#a-recommended-dev-flow-on-github-for-adding-new-models).
- [ ] [Galaxy quick](https://github.com/tenstorrent/tt-metal/actions/workflows/galaxy-quick.yaml) CI passes (if applicable)
- [ ] [Galaxy demo tests, for Llama](https://github.com/tenstorrent/tt-metal/actions/workflows/galaxy-demo-tests.yaml) CI passes, if applicable, because of current Llama work
- [ ] (For runtime and ops writers) [T3000 unit tests](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-unit-tests.yaml) CI passes (if applicable, since this is run on push to main)
- [ ] (For models and ops writers) [T3000 demo tests](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-demo-tests.yaml) CI passes (if applicable, since this is required for release)
- [ ] New/Existing tests provide coverage for changes

Aswinmcw committed:
Migrate rpow as llk op

Aswinmcw committed:
update

Aswinmcw committed:
update

Aswinmcw committed:
rebase

Aswinmcw committed:
rebase

Aswinmcw committed:
Refactor unary power operation tests and implementation

- Updated `test_unary_rpow_ttnn` to remove unused parameters and simplify input tensor creation.
- Introduced a new header file `ckernel_sfpu_rpow.h` for the rpow calculation logic.
- Modified `llk_math_eltwise_unary_sfpu_rpow` to utilize the new rpow calculation function.
- Adjusted the `rpow_tile` function to include destination accumulation mode.
- Included rpow header in `sfpu_split_includes.h` for better modularity.

Aswinmcw committed:
Remove rpow operation and related tests from the codebase

- Deleted the rpow operation implementation and its associated bindings in the unary composite operations.
- Updated the `test_unary_rpow_ttnn` to simplify the test case and remove unused assertions.
- Cleaned up commented-out code related to rpow in various files to improve code clarity.

Aswinmcw commented:
ULP graphs:
<img width="2805" height="2100" alt="existing" src="https://github.com/user-attachments/assets/fcfe2e12-9afb-45de-9c81-f89118d03891" />
<img width="2805" height="2100" alt="exp21f" src="https://github.com/user-attachments/assets/bfd4802c-a12d-44e5-bd8d-aa4351b65b73" />
<img width="2805" height="2100" alt="pow" src="https://github.com/user-attachments/assets/81fe7601-c378-41ab-99fc-cb46618ac297" />

Aswinmcw committed:
Merge branch 'main' into Aswinmcw/rpow_llk

Aswinmcw committed:
Remove duplicate case for CLAMP_TSS in is_parametrized_type function in unary_op_utils.hpp

Aswinmcw committed:
Update unary power operation with new parameterization and input range adjustments

### Changes made:
- Modified the `test_unary_rpow_ttnn` test to use a single exponent value of 5.0 and adjusted the input data range from [-20, 20] to [-30, 30].
- Updated the `calculate_rpow` function template to include a new boolean parameter for destination accumulation mode.
- Enhanced the initialization function for unary power operations to accommodate the new parameterization.

### Impact:
These changes improve the testing coverage and flexibility of the unary power operation, ensuring better performance and accuracy in computations.

Aswinmcw committed:
Merge branch 'main' into Aswinmcw/rpow_llk

Aswinmcw committed:
Refactor rpow implementation and update tests

### Changes
- Updated the `calculate_rpow` function to use `base_val` instead of `log_val` for clarity.
- Enhanced the `test_unary_rpow_ttnn` to include a broader range of exponent values and input shapes.
- Removed commented-out code and unnecessary print statements for cleaner test output.
- Adjusted documentation for the `rpow` function in the Python bindings to reflect the changes.

### Impact
These changes improve the clarity and functionality of the rpow operation, ensuring better test coverage and maintainability.

Aswinmcw committed:
Refactor `calculate_rpow` function for improved clarity

### Changes
- Simplified the implementation by directly using `sfpi::dst_reg[0]` in the power calculation, removing the intermediate variable.
- This change enhances readability and maintains the functionality of the `calculate_rpow` function across both blackhole and wormhole_b0 kernels.

### Impact
The refactor improves code clarity without altering the existing behavior of the rpow operation.

mcw-anasuya commented on tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_sfpu/ckernel_sfpu_rpow.h:
```diff
@@ -0,0 +1,26 @@
+// SPDX-FileCopyrightText: © 2023 Tenstorrent Inc.
```
`2025 Tenstorrent AI ULC`

Aswinmcw committed:
Merge branch 'main' into Aswinmcw/rpow_llk

Aswinmcw committed:
Update copyright year in header files from 2023 to 2025 for consistency across multiple kernel and math files.

Aswinmcw committed:
Merge branch 'main' into Aswinmcw/rpow_llk

Aswinmcw committed:
Merge branch 'main' into Aswinmcw/rpow_llk

Aswinmcw committed:
Merge branch 'main' into Aswinmcw/rpow_llk

Aswinmcw committed:
update

Aswinmcw committed:
update

KalaivaniMCW commented on tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_sfpu/ckernel_sfpu_rpow.h:
```diff
@@ -0,0 +1,23 @@
+namespace ckernel::sfpu {
+
+template <bool APPROXIMATION_MODE, int ITERATIONS, bool is_fp32_dest_acc_en>
+inline void calculate_rpow(const uint32_t base_val) {
```
pls add formula 
specifying base and exponent

Aswinmcw committed:
Update ULP threshold in unary power test and add comment for rpow function in kernel headers

Aswinmcw commented on tt_metal/hw/ckernels/wormhole_b0/metal/llk_api/llk_sfpu/ckernel_sfpu_rpow.h:
```diff
@@ -0,0 +1,23 @@
+template <bool APPROXIMATION_MODE, int ITERATIONS, bool is_fp32_dest_acc_en>
+inline void calculate_rpow(const uint32_t base_val) {
+    sfpi::vFloat base_val_v = Converter::as_float(base_val);
+#pragma GCC unroll 8
```
test working for larger shapes without any issues

KalaivaniMCW commented on tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_sfpu/ckernel_sfpu_rpow.h:
```diff
@@ -0,0 +1,23 @@
+#include "sfpi.h"
+
+namespace ckernel::sfpu {
+// ttnn.rpow(base, exponent) = pow(exponent, base)
```
*ttnn.rpow(exponent, scalar_base)

Aswinmcw committed:
Merge branch 'main' into Aswinmcw/rpow_llk

Aswinmcw committed:
Refactor rpow API documentation and implementation for clarity

Updated comments in rpow-related headers to clarify the function signature and its behavior. The change specifies that `ttnn.rpow(exponent, scalar_base)` computes `pow(scalar_base, exponent)`, enhancing readability and understanding of the API. Additionally, removed redundant namespace qualifiers in function calls for consistency across the codebase.

Aswinmcw committed:
Merge branch 'main' into Aswinmcw/rpow_llk
